### PR TITLE
bpo-45219: Factor dictkey indexing

### PR DESCRIPTION
**Pull request by markshannon**
Provides a better interface for querying dictkey objects and uses it `specialize.c`.

We will need this improved interface for future optimizations of object layout, so we might as well add it now.

Also fixes a super-obscure bug. If a globals dictionary contained a mutable key (your not supposed to, but the VM cannot prevent it), then a specialized `LOAD_GLOBAL` would not see the change in behavior when the key mutated.

https://bugs.python.org/issue45219

### CODE REVIEW

File: Objects/dictobject.c [hunk 1]
```diff
@@ -734,6 +734,64 @@ lookdict_index(PyDictKeysObject *k, Py_hash_t hash, Py_ssize_t index)
...
+    PyDictKeyEntry *ep0 = DK_ENTRIES(dk);
+    size_t mask = DK_MASK(dk);
+    Py_hash_t hash = PyObject_Hash(key);
+    assert (hash != -1);
+    size_t perturb = hash;
+    size_t i = (size_t)hash & mask;
+    Py_ssize_t ix;
+
+    /* Strings only */
+    for (;;) {
```
Comment (Fidget-Spinner) [inline Objects/dictobject.c:745]: Looks like some potential for factoring here. As far as I can tell this has a few lines different from the fast path for strings in `_Py_dict_lookup`.

File: Python/specialize.c [hunk 2]
```diff
@@ -1020,17 +1016,16 @@ _Py_Specialize_LoadGlobal(
...
     }
-    PyObject *value = NULL;
-    Py_ssize_t index = _PyDict_GetItemHint((PyDictObject *)globals, name, -1, &value);
-    assert (index != DKIX_ERROR);
     if (index != DKIX_EMPTY) {
         if (index != (uint16_t)index) {
             goto fail;
         }
-        uint32_t keys_version = _PyDictKeys_GetVersionForCurrentState((PyDictObject *)globals);
+        uint32_t keys_version = _PyDictKeys_GetVersionForCurrentState(((PyDictObject *)globals)->ma_keys);
```
Comment (Fidget-Spinner) [inline Python/specialize.c]: Line seems to be too long now. Same for the other times `_PyDictKeys_GetVersionForCurrentState` is used below.

**Review by Fidget-Spinner — COMMENTED** [review]
Thanks. This LGTM. I just have some stylistic suggestions.

File: Objects/dictobject.c [hunk 3]
```diff
@@ -734,6 +734,64 @@ lookdict_index(PyDictKeysObject *k, Py_hash_t hash, Py_ssize_t index)
...
+Py_ssize_t
+_PyDictKeys_StringLookup(PyDictKeysObject* dk, PyObject *key)
+{
+    DictKeysKind kind = dk->dk_kind;
+    if (!PyUnicode_CheckExact(key) || kind == DICT_KEYS_GENERAL) {
+        return DKIX_ERROR;
+    }
+    PyDictKeyEntry *ep0 = DK_ENTRIES(dk);
+    size_t mask = DK_MASK(dk);
+    Py_hash_t hash = PyObject_Hash(key);
```
Comment (methane) [inline Objects/dictobject.c]: How about this?

```c
    if ((hash = ((PyASCIIObject *) key)->hash) == -1) {
        hash = PyObject_Hash(key);  // or hash = PyUnicode_Type.tp_hash(key);
        if (hash == -1) {
            PyErr_Clear();
            return NULL;
        }
    }
```

**Pull request merged**